reader-privacy.md

```diff
@@ -0,0 +1,199 @@
+`hash(multihash)` that is used for IPNI lookups. Right now advertisments are not encrypted, but authenticated and contain plain multihash values in them. 
+That is going to change once *Writer Privacy* is implemented. Until then, an attacker could build a map of `hash(multihash) -> multihash` 
+by re-ingesting advertisements chain from each publisher in order to collect all original multihashes which can then be used to decrypt provider records and so on. 
+Doing that will require significant resources as it involves crawling the entire network. However, it will eventually be eliminated by *Writer Privacy* upgrade.
+
+Reader Privacy is a first step towards fully private content routing protocol. 
+
+Wider security implications are discussed in the IPFS Reader Privacy specification: TODO link here.
+
+#### Hashing and Encryption Function Upgrades
+
+All multihashes have a codec encoded in them. If a hashing or encryption funciton will have to rotate then different types of multihahses can coexist together
+and can be processed differently by IPNI implementations. It won't be possible to apply a fix retroactivelly to the data returned by previous lookup requests, 
+however IPNI implementations should start blocking all new ones that use a compromised scheme.
+
+Moving an IPNI implementation to a new hash / encryption function will involve reingesting all data from a scratch. Before Writer Privacy is impemented the 
+index can be migrated over to new functions by reingesting all advertisement chains. With Writer Privacy, Publishers will have to republish advertisments 
+using new algorithms. Both old and new scheme can coexist together for some time. The old one should be retired either immediately or once
+the indexes have been rebuilt and the users have been migrated over. 
+
+Exact operation procedure will be different for differnet IPNI implementations.
+
+### Trade Offs 
+
+* **Multiple lookups**. In the simplest scenario Reader Privacy protocol will require at least two roundtrips to find provider details for a given CID.
+It can be reduced down to one by caching `ProviderRecord`s locally at the client side. That would eliminate a need in a lookup
+per decrypted `ProviderRecordKey`. In the future there can be a separate service that distributes `PeerID` to `Multiaddresses` mappings in open.
+That dataset can be periodically downloaded by all clients and cached locally;
+
+* **Extra compute**. At minimum, clients will require to perform an extra hashing per CID and decryption per `ProviderRecordKey` that will add
+some overhead to each lookup. That overhead can be initially offloaded to the server but will have to be done by clients eventually (explained below);
+
+* **Extra storage space**. Storing encrypted data will require more space due to paddings and nonce;
+
+* **Bulk deletes**. Encrypted `PeerID` will be different for each multihash and hence bulk delete operations (delete everything for a provider X) will not be possible.
```

Comment:
> Encrypted `PeerID`

What is the purpose of encrypting the `PeerID`?

Reply:
On ingestion each multihash is going to be hashed and each peerID is going to be encrypted with a key derived from the multihash value. IPNI will store `hash(multihash) -> enc(peerID, multihash)` records so that it can return an encrypted peerID to a hashed lookup request. Otherwise it will have to return a raw peerID value that would leak identity of the provider to a passive observer.